### PR TITLE
Teachable learnset helper mechanics

### LINKED ISSUE #1379: Alolan Forms are missing Move Tutor learnsets
title.

### LINKED ISSUE #2274: Extend compatibility in teachable_learnsets.h
With the restructuring of how move compatibility is stored and moves no longer being linked to specific TMs, it is now possible to extend these teachable learnsets to include moves that are tutorable either by move tutor or by TM/TR in Generations IV-VIII (and potentially also Generations I-II, as well as FRLG). This eliminates the need for external tools such as PoryMoves (which for instance don't natively run on OSX) and allows people to add new move tutors and TMs without having to (manually) add compatibility for over 1,000 species. The only possible side effect is increased ROM space, which I personally believe is a fair sacrifice in function of ease of use.

### PR DESCRIPTION
**Pull request by Bassoonian**
This tool, which runs when building a ROM, does the following:
- Scans the repo for TMs and tutor moves
- Checks json files (from PoryMoves, with minor label adjustments for an easier time) for compatibility
- Creates a `teachable_learnsets.h` file with the smallest size possible by only including the moves found in the scan

Other features:
- Users can easily delete/add jsons for less/more data
- Any changes in `teachable_learnsets.h` that were made before this tool runs for the first time will be saved in `custom.json` for flawless migration

Fixes #1379 
Fixes #2274

## Internal to do list
- [x] list moves found in scan
- [x] integrate in Makefile
- [x] clarify python instructions in Install.md
- [x] update sv.json based on The Indigo Disk data ([source](https://gist.githubusercontent.com/SciresM/a1a0c2e48ae5ca58a0ecd131bb7f7845/raw/f7d5d4fc4c0de9420030421aa8d8b64fc5dbca22/personal.txt))

## Necessities before merging
- [x] #4026 needs to be merged into upcoming
- [x] #4027 needs to be resolved and (after master) be merged into upcoming

## What this PR does not contain
- XD data is currently limited to Deoxys' Normal form (as it contains compatibilities missing from all subsequent games)

## **Discord contact info**
bassoonian

### CODE REVIEW

File: tools/teachablelearnsethelper/teachable_learnset_helper.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+# scan incs
+incs_to_check =  glob.glob('./data/scripts/*.inc') # all .incs in the script folder
+incs_to_check += glob.glob('./data/maps/*/scripts.inc') # all map scripts
+for file in incs_to_check:
+    with open(file, 'r') as f2:
+        raw = f2.read()
+    if 'special ChooseMonForMoveTutor' in raw:
+        for x in re.findall('setvar VAR_0x8005, (MOVE_.*)', raw):
+            if not x in tutor_moves:
+                tutor_moves.append(x)
```
Comment (mrgriffin) [inline tools/teachablelearnsethelper/teachable_learnset_helper.py:20]: I think there's a few ways in which is this a little weird:
1. Somebody might use `setvar VAR_0x8005, MOVE_...` for something other than `ChooseMonForMoveTutor`, e.g. they might have a special that teaches a particular move (which is in `VAR_0x8005`, a sensible choice because `VAR_0x8004` contains the result of `ChoosePartyMon`)..
2. If somebody did, e.g. `copyvar VAR_0x8005, VAR_MOVE_TO_TEACH` then it won't be detected (and they'd probably be doing that in a way that makes it difficult to know what `VAR_MOVE_TO_TEACH` contains, at the very least through a `call`, but it could be a variable set in an entirely different script). Related, (and somewhat more likely) somebody could write a special which writes the move ID into `gSpecialVars_0x8005`.
3. If you write `@ setvar VAR_0x8005, MOVE_TACKLE` I think that would be picked up. This is both good and bad, bad because commenting things out should work, but good because it gives a way to explicitly list teachable moves in cases where it can't be picked up automatically (like 2).
4. Very, very minor, but this will find scripts that aren't OW scripts in `data/scripts` (i.e. not referenced by `data/event_scripts.s`). This is very unlikely to matter unless people write other script languages that can also reference OW vars.
5. If somebody happened to write move tutors in a macro they won't be detected. (I personally think a macro is a bad idea because I want to see what's happening to the special vars, but I could see other people thinking differently and could easily see a change like that making it into here or pret\*).

I don't have a workable suggestion for these because I think you'd have to do static analysis to work out which value is in `VAR_0x8005` when `special ChooseMonForMoveTutor` runs. In practice it probably won't come up except for people who are really pushing the boundaries of the scripting language. afaict MoT wouldn't see any bugs, although we do hit 1-3 points above and almost 4 (we just happen not to ever assign a move to `VAR_0x8005`).

tl;dr not using a parser, but I don't expect it to cause a problem. It's worth noting that something similar to 2 exists in vanilla: [`GetBattleFrontierTutorMoveIndex`](<https://github.com/pret/pokeemerald/blob/6c96641035099db8a2c1e680fb8793f1b29cf2fd/src/field_specials.c#L3188>) (it generates an index rather than a move ID because vanilla's tutors take indices not move IDs).

btw, I'd like to be really explicit that I love the idea of determining the smallest possible data structures at build time.

\* Not pret for specifically move tutors, because they work differently there.

**Comment by GraionDilach** [comment]
> Creates a teachable_learnsets.h file with the smallest size possible by only including the moves found in the scan

Not necessarily, because the generated header lacks the species_enabled gates. This means that with enough disabled species, the currently shipped alternative will be smaller than the generated result.

**Comment by Bassoonian** [comment]
> > Creates a teachable_learnsets.h file with the smallest size possible by only including the moves found in the scan
> 
> Not necessarily, because the generated header lacks the species_enabled gates. This means that with enough disabled species, the currently shipped alternative will be smaller than the generated result.

I'm unsure what you mean. The file generated preserves the P_FAMILY checks, so even if the generated result would be larger than the currently shipped .h file (which is technically _too_ small because it lacks some learnsets and compatibilities), all the disabled species wouldn't be compiled in the game. The final result is still the smallest possible list for each species (with only the enabled species being compiled, just as the vanilla file).

**Comment by GraionDilach** [comment]
Hm, my python knowledge failed here. I was quite certain this strips away the preprocessor lines, but my test show it doesn't. I don't understand that part, but okay.

**Comment by Bassoonian** [comment]
It simply regexes the existing `static const u16 s<speciesnamehere>TeachableLearnset[]`` without touching what actually surrounds them, so the family defines (and comments, if the user would for some reason want to add comments to the file, and #defines, etc) are preserved in their original position

**Comment by Bassoonian** [comment]
This PR is now ready for review. Please make sure to check the Python instructions in the installation instructions properly, as I have very little (read: no) experience outside of OSX, and even on my own computer I forgot how I installed it :)

**Comment by AlexOn1ine** [comment]
Is there a point in adding a `custom.json` file that has only the species data without the moves for users so they know they can add new moves to it?

**Comment by Bassoonian** [comment]
I don't think so, considering the thing will make a `custom.json` for you if there's any non-vanilla data in the file when you first run it. You can also name it whatever as long as it's a json file, so there could be a `gaia.json` or an `unbound.json` or whatever you'd prefer.

**Comment by AsparagusEduardo** [comment]
I tested by removing this:
image

And now it's not building, with no information about the error:
image

Also, I think it'd be better to have the "Updated" message formatted like this:
```
Updated Teachable Learnsets for: Charmander, Charmeleon, Charizard, Squirtle, ...
```

**Comment by Bassoonian** [comment]
Oh, huh, interesting. It's refusing to print anything when triggered by Makefile on my end, while it seems to ignore newlines in yours. I think that the error is quite literally what it says, i.e. the print is being considered a command by bash? Do you get the same error if you just run it without removing the Dynamic Punch line?

**Comment by AsparagusEduardo** [comment]
> Do you get the same error if you just run it without removing the Dynamic Punch line?
I first ran make with the branch as-is. I removed the line afterwards and did make.

Restoring the line doesn't fix it.
image

**Comment by AsparagusEduardo** [comment]
> Oh, huh, interesting. It's refusing to print anything when triggered by Makefile on my end, while it seems to ignore newlines in yours. I think that the error is quite literally what it says, i.e. the print is being considered a command by bash?

I think it'll be better to just separate this as its own tool instead of having it part of makefile, which was something I requested previously.

**Comment by Bassoonian** [comment]
We previously reached a majority vote on just integrating it, since if it's not built into the makefile it may just as well not be a part of the repo. It worked previously, I don't know what broke it, so I'll try to fix it

**Comment by AsparagusEduardo** [comment]
> We previously reached a majority vote on just integrating it, since if it's not built into the makefile it may just as well not be a part of the repo. It worked previously, I don't know what broke it, so I'll try to fix it

Where was the vote?

File: Makefile [hunk 2]
```diff
@@ -461,6 +462,10 @@ $(OBJ_DIR)/sym_common.ld: sym_common.txt $(C_OBJS) $(wildcard common_syms/*.txt)
 $(OBJ_DIR)/sym_ewram.ld: sym_ewram.txt
 	$(RAMSCRGEN) ewram_data $< ENGLISH > $@
 
+# NOTE: Depending on event_scripts.o is hacky, but we want to depend on everything event_scripts.s depends on without having to alter scaninc
+$(DATA_SRC_SUBDIR)/pokemon/teachable_learnsets.h: $(DATA_ASM_BUILDDIR)/event_scripts.o
+	$(TEACHABLELEARNSETHELPER)
```
[the remark this answers is not in the document]
Comment (Bassoonian) [reply Makefile:467]: This solved the issue indeed, both the (occasional lack of) printing and the printing being interpreted by bash. Thanks! :)

**Comment by Bassoonian** [comment]
Fixed Makefile (thanks MGriffin!), expanded the header to refer to the Python file and added a config to easily disable the helper without having to fiddle with Makefile. Ready for re-review!

File: include/config/pokemon.h [hunk 3]
```diff
@@ -37,6 +37,9 @@
 #define P_SHOW_TERA_TYPE            GEN_LATEST  // Since Gen 9, the Tera Type is shown on the summary screen.
 #define P_TM_LITERACY               GEN_LATEST  // Since Gen 6, TM illiterate Pokémon can learn TMs that teach moves that are in their level-up learnsets.
 
+// Learnset helper toggles
+#define P_LEARNSET_HELPER_TEACHABLE TRUE        // If TRUE, teachable_learnsets.h will be populated by tools/learnset_helpers/teachable.py using the included json files based on available TMs and tutors.
```
Comment (AsparagusEduardo) [inline include/config/pokemon.h:41]:
```suggestion
#define P_LEARNSET_HELPER_TEACHABLE TRUE        // If TRUE, teachable_learnsets.h will be populated by tools/learnset_helpers/teachable.py using the included JSON files based on available TMs and tutors.
```

File: tools/learnset_helpers/teachable.py [hunk 4]
```diff
@@ -3,6 +3,11 @@
 import json
 import os
 
+# before all else, abort if the config is off
+with open("./include/config/pokemon.h", "r") as file:
+    if re.findall("#define P_LEARNSET_HELPER_TEACHABLE *([^ ]*)", file.read())[0] != "TRUE":
+        quit()
```
Comment (AsparagusEduardo) [inline tools/learnset_helpers/teachable.py:9]: There's no handling in case `P_LEARNSET_HELPER_TEACHABLE` doesn't exist.
image

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]
We're almost there

File: tools/learnset_helpers/teachable.py (hunk 4 above)
Comment (Bassoonian) [reply tools/learnset_helpers/teachable.py:9]: Good catch!

**Review by AsparagusEduardo — APPROVED** [review]
